### PR TITLE
ramips: add support for TP-Link RE305 v3

### PR DESCRIPTION
**Pull request by koyuch**
Specs (same as in v1):
- MT7628AN (575 MHz)
- 64MB RAM
- 8MB of flash (SPI NOR)
- 1x 10/100Mbps Ethernet (MT7628AN built-in switch with vlan)
- 1x 2.4GHz wifi (MT7628AN)
- 1x 5Ghz wifi (MT7612E)
- 4x LEDs (5 GPIO-controlled)
- 1x reset button
- 1x WPS button

The only and important difference between v1 & v3 is in flash memory layout, so pls don't interchange these 2 builds!

So far verified way to load the build:
1. connect trough serial (1n8, baudrate=57600)
2. setup the TFTP server and connect it via ethernet (ipaddr=192.168.0.254 of device, serverip=192.168.0.184 - your pc)
3. boot from a initramfs image first (choose 1 in the bootloader options)
4. test it a bit with that, then proceed to run sysupgrade build

Signed-off-by: servitkar <servitkar[at]gmail[dot]com>

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -30,3 +30,4 @@ git-src
 .cproject
 .ccache
 .vscode
+.idea
```
Comment (hauke) [inline .gitignore:33]: Please add this in a separate pull request as this is unrelated to the other changes.

Comment (koyuch) [reply .gitignore]: ok. done :+1:

**Comment by adschm** [comment]
This has formal issues. Please visit https://openwrt.org/submitting-patches another time.

**Comment by koyuch** [comment]
> This has formal issues. Please visit https://openwrt.org/submitting-patches another time.
> 
> Missing commit message, fix Signed-off-by, ...

ah ok. I thought that merged branch is squashed & PR's description will become commit's description (this is how we work on my current project..)
@adschm is it better now?

**Comment by koyuch** [comment]
or do you need there my real name..? (in the signed-off?)

**Comment by koyuch** [comment]
should I change also commit's author name..?

**Comment by adschm** [comment]
We need your real name and working e-mail address for both author and Signed-off-by.

**Comment by adschm** [comment]
> > This has formal issues. Please visit https://openwrt.org/submitting-patches another time.
> > Missing commit message, fix Signed-off-by, ...
> 
> ah ok. I thought that merged branch is squashed & PR's description will become commit's description (this is how we work on my current project..)
> @adschm is it better now?

Only the information in the commit will be kept.

**Comment by koyuch** [comment]
isn't "real" email spam-prone..? (btw. it's "working" email..)

**Comment by koyuch** [comment]
@adschm better now?

**Comment by adschm** [comment]
> @adschm better now?

Yes, thanks. This this is very similar to re305-v1, as you state yourself, it would be a good idea to create a shared DTSI for both of them. The DTS files would then only contain model/compatible, the spi0 node, and the MAC address assignments.

You should find various examples in the ramips DTS folder...

**Comment by adschm** [comment]
And please wrap your commit message, so lines have max. 74 characters.

**Comment by koyuch** [comment]
@adschm yes I've seen some dtsi files, but can I really use so much in common? because flash layout is different (config partition is on different position & has different length). I think the only thing left in common are gpios.. (thats why I havent tried to create dtsi so far..)

**Comment by adschm** [comment]
Well, keys, leds, and everything from pcie etc. that is not mac-address-related ...

**Comment by Borromini** [comment]
@koyuch Could you document the exact pin layout as well please? Where it's located on the board and what pin does what (VCC, Tx, Rx, GND)?

Just got a hand-me-down and opened it up, I see a J1 header on a daughter board behind the wall socket, but the wiki entry has no info on serial on the RE305 v1 either unfortunately, so little to go on.

**Comment by Borromini** [comment]
@koyuch The web UI upgrade issue can be fixed by setting a `.soft_ver` like this:

```
diff --git a/tools/firmware-utils/src/tplink-safeloader.c b/tools/firmware-utils/src/tplink-safeloader.c
index 626c9a434b..74acc631d5 100644
--- a/tools/firmware-utils/src/tplink-safeloader.c
+++ b/tools/firmware-utils/src/tplink-safeloader.c
@@ -2090,7 +2090,8 @@ static struct device_info boards[] = {
                        "{product_name:RE305,product_ver:3.0.0,special_id:43410000}\n"
                        "{product_name:RE305,product_ver:3.0.0,special_id:52550000}\n",
                .part_trail = 0x00,
-               .soft_ver = NULL,
+               /** Latest EU firmware is 1.1.3 Build 20200715 Rel. 38287 **/
+               .soft_ver = "soft_ver:2.0.0\n",
 
                .partitions = {
                        {"fs-uboot", 0x00000, 0x20000},
```

This is on top of your PR. Just apply & squash.

**Comment by koyuch** [comment]
@Borromini thanks for the update. I've added it here :) Still I need to resolve this dtsi thing..

**Comment by Borromini** [comment]
@koyuch This might not be the most orthodox way, but I didn't know any more convenient way to share it. So you can find the split out DTS set here:

- [DTSI for both RE305 models](https://paste.debian.net/1193411/)
- [RE305 v1 DTS](https://paste.debian.net/1193416/)
- [RE305 v3 DTS](https://paste.debian.net/1193417/)

The pastebin has a 'Download' button. I have compile-tested this on 21.02 HEAD (DTSes were built on master though), and run-tested as well.

I'm not sure if the license identifiers should be in every file in the dependency chain, but that will become clear upon review.

@adschm I remembered your question about the whole PCIE node from last time so I hope I've done it properly here (if not you'll surely tell me :smile: ).

**Comment by Borromini** [comment]
@koyuch Mind you, you need to include the **v1 DTS** as well, since that uses the new DTSI just like the v3 DTS does. If you don't you just break the RE305 v1.

**Comment by koyuch** [comment]
@Borromini I've had this prepared in the evening but felt asleep sooner than commited :D Is it like split acceptable/workable? especialy can you have common `partition@20000` & have specific partition sizes in concrete versions, or it doesn't make sense..? (I've compiled it but don't know if it would work..)
yy, I'll add also v1 split - I was "just playing" :D

**Comment by koyuch** [comment]
@Borromini I've added also your `wifi5` section. seems clever :) thanks :+1:

**Comment by Borromini** [comment]
Just use what I prepped please. It looks like you're splitting up the spi node as well, looks like a bad idea (if it even works). The wifi label is apparently the clean way to set/add stuff, instead of defining the whole node again.

Like I said, the files I linked to are compiled and run-tested on a v3.

**Comment by Borromini** [comment]
There seem to be end brackets missing in both the v1 and v3 DTS as suggested [in this forum post](https://forum.openwrt.org/t/support-for-tp-link-re305-v3/75893/29).

**Comment by adschm** [comment]
Please squash your changes into a single proper commit.

And as the last commit shows, you should test your changes.

**Comment by koyuch** [comment]
ok, I'll do that. funny thing is, that it actually did compile on my env even without brackets..

**Comment by adschm** [comment]
Ping.

**Comment by adschm** [comment]
> ok, I'll do that. funny thing is, that it actually did compile on my env even without brackets..

That's not really helpful for building trust in your testing ...

**Comment by thecoder-001** [comment]
Have been testing latest OpenWRT compiled with the PR for a few days and everything seems to be working pretty well and looks to be stable so far. Others too have reported success on the earlier mentioned [openwrt forum post](https://forum.openwrt.org/t/support-for-tp-link-re305-v3/75893/39).

**Comment by koyuch** [comment]
> That's not really helpful for building trust in your testing ...
I just followed official guide on building openwrt..

OK, so I've squashed & fixed commit message. Build has been tested by more people from this PR. Is it now OK for you @adschm ?

**Review by Mauroben — APPROVED** [review]

**Comment by adschm** [comment]
Has anybody tested factory flashing via WebUI, so this can be provided with verified instructions?

**Comment by thecoder-001** [comment]
> Has anybody tested factory flashing via WebUI, so this can be provided with verified instructions?

I can confirm that factory flashing via WebUI works with no problems. At least [another person on the forum](https://forum.openwrt.org/t/support-for-tp-link-re305-v3/75893/38?u=thecoder-001) too was successful to factory flash.

**Comment by Mauroben** [comment]
> Has anybody tested factory flashing via WebUI, so this can be provided with verified instructions?

Yes, work fine for me.

**Comment by ingmar** [comment]
I'm also running this branch successfully. Installed via OEM WebUI from the factory.bin. Everything seems to work, including LEDs.

**Comment by adschm** [comment]
With so many testers, it shouldn't be hard for somebody to write proper flashing instructions and post them here.

**Comment by ingmar** [comment]
Something like this?

To switch the RE305 v3 from OEM (factory) firmware to OpenWRT, follow these steps:

- Visit http://tplinkrepeater.net or the configured IP address of your RE305 v3 (default [192.168.0.254](http://192.168.0.254)).
- Log in with the password you've set during initial setup of the RE305 (there is no default password).
- Go to **Settings** -> **System Tools** -> **Firmware upgrade**
- Click **Browse** and select the OpenWRT _factory.bin_ (not sysupgrade.bin)
- A window with a progress bar will appear. Wait until it completes.
- The RE305 will reboot into OpenWRT and serve DHCP requests on the ethernet port.
- Connect an RJ45 cable from the RE305 to your computer and access LuCI at http://192.168.1.10/cgi-bin/luci/ to configure (or use ssh).

**Comment by nanomad** [comment]
> * Connect an RJ45 cable from the RE305 to your computer and access LuCI at http://192.168.1.10/cgi-bin/luci/ to configure (or use ssh).

Just tested and it works perfectly, but the default IP is 1.1 and not 1.10

**Comment by ingmar** [comment]
> > * Connect an RJ45 cable from the RE305 to your computer and access LuCI at http://192.168.1.10/cgi-bin/luci/ to configure (or use ssh).
> 
> Just tested and it works perfectly, but the default IP is 1.1 and not 1.10

Thanks, I've gone and fixed that in the above comment.

**Comment by koyuch** [comment]
OK, I've updated the commit message, according to your steps..

**Comment by koyuch** [comment]
There have been conflicts opposed to original master, so I've had to rebase it & change it accordingly for v3 config. (There are other namings like nvmem-cells instead of mtd-mac-address etc..). It has to be tested properly though..

**Comment by nanomad** [comment]
> There have been conflicts opposed to original master, so I've had to rebase it & change it accordingly for v3 config. (There are other namings like nvmem-cells instead of mtd-mac-address etc..). It has to be tested properly though..

I've flashed it on my RE305 v3 (upgrading before your force pull) and it seems to be working just fine. I've noticed that the WLAN mac address changed though (last bit is 0 instead of 1, while 1 is now assigned to the eth interface)

**Comment by adschm** [comment]
I rearranged some stuff and put the result here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/re305

The only thing missing IMO is the correct MAC address assignment for the v3. Can somebody please post a table like described here, so we can choose the correct addresses and get this done:

https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

**Comment by adschm** [comment]
Nobody?

**Comment by koyuch** [comment]
@adschm ok, so should I force push here your version? (I don't know what are your changes to it though..)

to MAC addresses - this is issue we were solving in the beggining with pmeira https://forum.openwrt.org/t/support-for-tp-link-re305-v3/75893/7 - I was checking it with my (in that time oem firmware & it was really these negative increments (I would bet, that also v1 would have those..) Though it would be nice if someone with OEM could double check this..

**Comment by adschm** [comment]
> ok, so should I force push here your version? (I don't know what are your changes to it though..)

No, please simply do what I asked for in my previous post, and provide the table of MAC addresses based on the description in the link.

File: tools/firmware-utils/src/tplink-safeloader.c [hunk 2]
```diff
@@ -2207,6 +2207,47 @@ static struct device_info boards[] = {
...
+			"{product_name:RE305,product_ver:3.0.0,special_id:4B520000}\n"
+			"{product_name:RE305,product_ver:3.0.0,special_id:41550000}\n"
+			"{product_name:RE305,product_ver:3.0.0,special_id:42520000}\n"
+			"{product_name:RE305,product_ver:3.0.0,special_id:55530000}\n"
+			"{product_name:RE305,product_ver:3.0.0,special_id:45530000}\n"
+			"{product_name:RE305,product_ver:3.0.0,special_id:41530000}\n"
+			"{product_name:RE305,product_ver:3.0.0,special_id:43410000}\n"
+			"{product_name:RE305,product_ver:3.0.0,special_id:52550000}\n",
+		.part_trail = 0x00,
+		.soft_ver = "soft_ver:2.0.0\n",
```
Comment (Borromini) [inline tools/firmware-utils/src/tplink-safeloader.c:2228]: This needs adapting after 8dba4741b3 - should become:

`.soft_ver = SOFT_VER_TEXT("soft_ver:2.0.0\n"),`

**Comment by d-me3** [comment]
> I rearranged some stuff and put the result here: https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/re305
> 
> The only thing missing IMO is the correct MAC address assignment for the v3. Can somebody please post a table like described here, so we can choose the correct addresses and get this done:
> 
> https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

I just received my RE305 v3 today. Below are the MAC addresses from the stock firmware (I've redacted one portion for privacy reasons but I believe it is sufficient for understanding the MAC addresses order). Please let me know if this is sufficient.

Thanks!

5.0 Ghz: 00-31-92-XX-7F-D0
2.4 Ghz: 00-31-92-XX-7F-D1
Ethernet: 00-31-92-XX-7F-D2

**Comment by d-me3** [comment]
> I rearranged some stuff and put the result here: https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/re305
> 
> The only thing missing IMO is the correct MAC address assignment for the v3. Can somebody please post a table like described here, so we can choose the correct addresses and get this done:
> 
> https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

@adschm I just installed OpenWRT on RE305 v3, and in fact all MAC addresses (eht0, wlan0 and wlan1) are showing the same:

```
root@OpenWrt:~# cat /sys/class/net/eth0/address
00:31:92:xx:7f:d2 
root@OpenWrt:~# cat /sys/class/net/wlan0/address
00:31:92:xx:7f:d2
root@OpenWrt:~# cat /sys/class/net/wlan1/address
00:31:92:xx:7f:d2
root@OpenWrt:~#
```

eth0 is correct, but wlan0 and wlan1 are wrong.

wlan0 (2ghz) should be (eth0-1) and wlan1 (5ghz) should be (eth0-2).

Is there any additional information you need I can get from command line via SSH?

**Comment by d-me3** [comment]
Confirmed. I just merged the green lines above from @koyuch into @adschm's  "mt7628an_tplink_re305.dtsi" and now all MAC addresses are OK, the same as the stock firmware:

```
root@OpenWrt:~# ip link | awk '{print $2}'
lo:
00:00:00:00:00:00
eth0:
00:31:92:xx:7f:d2
br-lan:
00:31:92:xx:7f:d2
wlan0:
00:31:92:xx:7f:d1
wlan1:
00:31:92:xx:7f:d0
root@OpenWrt:~#
```

**Comment by koyuch** [comment]
ok, so I've added that comment to commit, rebased on master.. Just wondering where did `tplink-safeloader.c` go.. are utils now in some separate repo..?

**Comment by koyuch** [comment]
@adschm I've included your patches into this PR. the only missing thing were those increments.. is it ok now? & what about those utils..?

**Review by dmsza — APPROVED** [review]

**Comment by adschm** [comment]
I've taken of the firmware-utils changes myself; I will do a final test-build and then merge this with the proper MAC addresses afterwards. Thanks for keeping up.

**Comment by koyuch** [comment]
@adschm thanks. btw. in `tplink-safeloader.c` I would suggest what @Borromini has proposed before ^^^ - soft version definition like this:
> `.soft_ver = SOFT_VER_TEXT("soft_ver:2.0.0\n"),`

**Comment by adschm** [comment]
> @adschm thanks. btw. in `tplink-safeloader.c` I would suggest what @Borromini has proposed before ^^^ - soft version definition like this:
> 
> > `.soft_ver = SOFT_VER_TEXT("soft_ver:2.0.0\n"),`

Yes, I already picked that up and forgot to mention it.
